File: windows/src/AirClip.Crypto/PairingInvite.cs
using System.Text;

namespace AirClip.Crypto;

/// <summary>
/// The payload behind a pairing QR code: <c>airclip://pair?v=1&amp;k=…&amp;n=…&amp;s=…&amp;p=…</c>. The phone scans
/// it, joins the group and knows where to look, all without anyone typing thirty-two characters.
/// <para>
/// <see cref="ToString"/> deliberately does <em>not</em> return the URI. The URI contains the group
/// secret, and a type whose default string form is a secret ends up in a log file sooner or later; call
/// <see cref="ToUri"/> when the secret is genuinely what is wanted.
/// </para>
/// </summary>
public sealed class PairingInvite
{
    public const string Scheme = "airclip";
    public const string PairHost = "pair";
    public const string Version = "1";

    public PairingInvite(PairingKey key, string deviceName, string serviceName, int port)
    {
        ArgumentNullException.ThrowIfNull(key);
        Key = key;
        DeviceName = string.IsNullOrWhiteSpace(deviceName) ? "AirClip" : deviceName.Trim();
        ServiceName = string.IsNullOrWhiteSpace(serviceName) ? "_airclip._tcp.local." : serviceName.Trim();
        Port = port is >= 1 and <= 65535 ? port : 47653;
    }

    public PairingKey Key { get; }

    public string DeviceName { get; }

    public string ServiceName { get; }

    public int Port { get; }

    public Uri ToUri()
    {
        var builder = new StringBuilder($"{Scheme}://{PairHost}?v={Version}");
        builder.Append("&k=").Append(Uri.EscapeDataString(Key.Code));
        builder.Append("&n=").Append(Uri.EscapeDataString(DeviceName));
        builder.Append("&s=").Append(Uri.EscapeDataString(ServiceName));
        builder.Append("&p=").Append(Port);
        return new Uri(builder.ToString());
    }

    public static bool TryParse(string? text, out PairingInvite? invite)
    {
        invite = null;
        if (string.IsNullOrWhiteSpace(text)
            || !Uri.TryCreate(text.Trim(), UriKind.Absolute, out Uri? uri)
            || !string.Equals(uri.Scheme, Scheme, StringComparison.OrdinalIgnoreCase)
            || !string.Equals(uri.Host, PairHost, StringComparison.OrdinalIgnoreCase))
        {
            return false;
        }

        Dictionary<string, string> query = ParseQuery(uri.Query);
        if (!query.TryGetValue("k", out string? code) || !PairingKey.TryParse(code, out PairingKey? key))
        {
            return false;
        }

        query.TryGetValue("n", out string? name);
        query.TryGetValue("s", out string? service);
        int port = query.TryGetValue("p", out string? rawPort) && int.TryParse(rawPort, out int parsed)
            ? parsed
            : 47653;

        invite = new PairingInvite(key!, name ?? "AirClip", service ?? "_airclip._tcp.local.", port);
        return true;
    }

    /// <summary>Safe to log: describes the invite by fingerprint and omits the secret entirely.</summary>
    public override string ToString() => $"配对邀请 {DeviceName} · {ServiceName} · :{Port} · {Key.Fingerprint}";

    private static Dictionary<string, string> ParseQuery(string query)
    {
        var values = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        foreach (string pair in query.TrimStart('?').Split('&', StringSplitOptions.RemoveEmptyEntries))
        {
            int split = pair.IndexOf('=', StringComparison.Ordinal);
            if (split > 0)
            {
                values[Uri.UnescapeDataString(pair[..split])] = Uri.UnescapeDataString(pair[(split + 1)..]);
            }
        }

        return values;
    }
}

File: android/app/src/main/kotlin/com/airclip/core/crypto/PairingInvite.kt
package com.airclip.core.crypto

import com.airclip.core.protocol.ProtocolConstants
import java.io.UnsupportedEncodingException
import java.net.URI
import java.net.URLDecoder
import java.net.URLEncoder

/**
 * The payload behind a pairing QR code: `airclip://pair?v=1&k=…&n=…&s=…&p=…`. The other device scans
 * it, joins the group and knows where to look, all without anyone typing thirty-two characters.
 *
 * A port of `AirClip.Crypto.PairingInvite`, down to the one-letter parameter names — a QR code is only
 * useful if the device on the other side of the camera reads the same field names this one writes.
 *
 * [toString] deliberately does *not* return the URI. The URI contains the group secret, and a type
 * whose default string form is a secret ends up in a log file sooner or later; call [toUri] when the
 * secret is genuinely what is wanted.
 *
 * Parsed with `java.net` rather than `android.net.Uri` to keep this class testable off-device.
 */
class PairingInvite(
    val key: PairingKey,
    deviceName: String,
    serviceName: String,
    port: Int,
) {
    val deviceName: String = deviceName.trim().ifBlank { "AirClip" }
    val serviceName: String = serviceName.trim().ifBlank { DEFAULT_SERVICE }
    val port: Int = if (port in 1..65535) port else ProtocolConstants.DEFAULT_PORT

    fun toUri(): String = buildString {
        append(SCHEME).append("://").append(PAIR_HOST).append("?v=").append(VERSION)
        append("&k=").append(encode(key.code))
        append("&n=").append(encode(deviceName))
        append("&s=").append(encode(serviceName))
        append("&p=").append(port)
    }

    /** Safe to log: describes the invite by fingerprint and omits the secret entirely. */
    override fun toString(): String =
        "配对邀请 $deviceName · $serviceName · :$port · ${key.fingerprint}"

    companion object {
        const val SCHEME = "airclip"
        const val PAIR_HOST = "pair"
        const val VERSION = "1"

        /** The full mDNS name. `NsdManager` wants the shorter `ProtocolConstants.SERVICE_TYPE`. */
        const val DEFAULT_SERVICE = "_airclip._tcp.local."

        /** `null` for anything that is not an `airclip://pair` URI carrying a usable pairing code. */
        fun parse(text: String): PairingInvite? {
            val uri = runCatching { URI(text.trim()) }.getOrNull() ?: return null
            if (!SCHEME.equals(uri.scheme, ignoreCase = true)) return null
            // Both airclip://pair?... and airclip:pair?... are accepted; QR scanners differ.
            val target = uri.authority ?: uri.path?.trim('/')
            if (!PAIR_HOST.equals(target, ignoreCase = true)) return null

            val query = parseQuery(uri.rawQuery ?: return null)
            val key = query["k"]?.let(PairingKey::parse) ?: return null
            val port = query["p"]?.toIntOrNull() ?: ProtocolConstants.DEFAULT_PORT

            return PairingInvite(
                key = key,
                deviceName = query["n"].orEmpty(),
                serviceName = query["s"].orEmpty(),
                port = port,
            )
        }

        private fun parseQuery(raw: String): Map<String, String> = raw.split('&')
            .mapNotNull { pair ->
                val index = pair.indexOf('=')
                if (index <= 0) null else decode(pair.substring(0, index)) to decode(pair.substring(index + 1))
            }
            .toMap()

        private fun decode(value: String): String = try {
            URLDecoder.decode(value, "UTF-8")
        } catch (e: UnsupportedEncodingException) {
            value
        } catch (e: IllegalArgumentException) {
            value
        }

        private fun encode(value: String): String = URLEncoder.encode(value, "UTF-8").replace("+", "%20")
    }
}
